room_allocation.py

```diff
@@ -79,21 +89,29 @@ def add_person(self, person_name, typ, wants_accomodation=False):
                 habitable_rooms.append(room)
 
         try:
-            room = random.choice(habitable_rooms)
-            room.people_in_room.append(person)
-            return person
+            if habitable_rooms:
+                room = random.choice(habitable_rooms)
+                room.people_in_room[person.name] = person
+                return person
+            else:
+                self.unallocated_people[person_name] = person
+                return "No room to add person to."
         except IndexError as error:
+            self.unallocated_people[person_name] = person
             message = ' (No room to add %s to.)' % (type(person).__name__)
-            raise type(error)  # (error.message + message)
+            # raise type(error)  (error.message + message)
+            return "No room to add person to."
 
     def can_be_in_room(self, person, room):
-        if person in room.people_in_room:
+        if person in room.people_in_room.values() or room.is_full():
             return False
         if type(person) is Fellow:
             if type(room) is Office:
                 return True
-            if type(room) is LivingSpace and person.wants_accomodation:
+            if type(room) is LivingSpace and person.wants_accommodation:
                 return True
+            else:
+                return False
         elif type(person) is Staff:
```

Comment:
You may as well join the two. and have `type(person) is Staff and type(room) is LivingSpace`